### ISSUE TITLE
Custom Mapped interface types fail with Exception

### ISSUE THREAD

**Issue by AMilassin** [vojtechhabarta/typescript-generator#439]
TypeScript generation fails for interface types when they are involved in CustomTypeMapping. See Exception and minimized test.

```
java.lang.IndexOutOfBoundsException: Index: 0

	at java.util.Collections$EmptyList.get(Collections.java:4454)
	at cz.habarta.typescript.generator.CustomMappingTypeProcessor.lambda$processType$2(CustomMappingTypeProcessor.java:39)
	at cz.habarta.typescript.generator.CustomMappingTypeProcessor.processType(CustomMappingTypeProcessor.java:50)
	at cz.habarta.typescript.generator.TypeProcessor$Chain.processType(TypeProcessor.java:104)
	at cz.habarta.typescript.generator.TypeProcessor.processTypeInTemporaryContext(TypeProcessor.java:21)
	at cz.habarta.typescript.generator.parser.ModelParser.parseQueue(ModelParser.java:93)
	at cz.habarta.typescript.generator.parser.ModelParser.parseModel(ModelParser.java:61)
	at cz.habarta.typescript.generator.TypeScriptGenerator.generateTypeScript(TypeScriptGenerator.java:82)
	at cz.habarta.typescript.generator.TypeScriptGenerator.generateTypeScript(TypeScriptGenerator.java:73)
	at cz.habarta.typescript.generator.TypeScriptGenerator.generateTypeScript(TypeScriptGenerator.java:68)
	at cz.habarta.typescript.generator.InterfaceTest.testGenericInterfaceWithCustomMapping(InterfaceTest.java:48)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

```

Minimal test to verify the issue (code added to `InterfaceTest.java`):
```
 
public class InterfaceTest {

    @Test
    public void test() {
        final Settings settings = TestUtils.settings();
        final String output = new TypeScriptGenerator(settings).generateTypeScript(Input.from(Book.class));
        Assert.assertTrue(output.contains("interface Book"));
        Assert.assertTrue(output.contains("title: string;"));
        Assert.assertTrue(output.contains("interface Author"));
        System.out.println(output);
    }

    @Test
    public void testReadOnlyProperties() {
        final Settings settings = TestUtils.settings();
        settings.declarePropertiesAsReadOnly = true;
        final String output = new TypeScriptGenerator(settings).generateTypeScript(Input.from(Book.class));
        Assert.assertTrue(output.contains("readonly author: Author;"));
        Assert.assertTrue(output.contains("readonly title: string;"));
        Assert.assertTrue(output.contains("readonly name: string;"));
    }

    // fails
    @Test
    public void testInterfaceWithCustomMapping() {
        final Settings settings = TestUtils.settings();
        settings.customTypeMappings = new HashMap<String, String>(){{
            put("cz.habarta.typescript.generator.InterfaceTest$Book", "CustomBook");
        }};
        final String output = new TypeScriptGenerator(settings).generateTypeScript(Input.from(Book.class));
        Assert.assertTrue(output.contains("interface CustomBook"));
    }

    // fails
    @Test
    public void testGenericInterfaceWithCustomMapping() {
        final Settings settings = TestUtils.settings();
        settings.customTypeMappings = new HashMap<String, String>(){{
            put("cz.habarta.typescript.generator.InterfaceTest$GenericInterface<D>", "CustomGenericInterface<D>");
        }};
        final String output = new TypeScriptGenerator(settings).generateTypeScript(Input.from(GenericInterface.class));
        System.out.println(output);
        Assert.assertTrue(output.contains("interface CustomGenericInterface<D>"));
    }

    static interface Book {
        Author getAuthor();
        String getTitle();
    }

    static interface Author {
        String getName();
    }

    static interface GenericInterface<D> {
        D getType();
    }

}

```

**Comment by vojtechhabarta** [comment]
Thanks for this bug report with example test. I realized that this case fails not only with interfaces but also with classes. I fixed it in 5f919da8e879f7795ca3b94dc3cba4cec2d6c07e.

I think that instead of [customTypeMappings](http://www.habarta.cz/typescript-generator/maven/typescript-generator-maven-plugin/generate-mojo.html#customTypeMappings) you need [customTypeNaming](http://www.habarta.cz/typescript-generator/maven/typescript-generator-maven-plugin/generate-mojo.html#customTypeNaming) but that's just my guess from your tests.

**Comment by vojtechhabarta** [comment]
Released [2.19.575](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.19.575) with the fix.
